{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "29-Apr-2025 12:54:01 DEBUG utils - Directory ./ already exists. Skipping.\n",
      "29-Apr-2025 12:54:01 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE206nnn/GSE206762/soft/GSE206762_family.soft.gz to ./GSE206762_family.soft.gz\n",
      "100%|██████████| 3.41k/3.41k [00:00<00:00, 108kB/s]\n",
      "29-Apr-2025 12:54:02 DEBUG downloader - Size validation passed\n",
      "29-Apr-2025 12:54:02 DEBUG downloader - Moving /tmp/tmplwv2f77u to /cis/home/iessien1/Documents/injury_atlas/GSE206762_family.soft.gz\n",
      "29-Apr-2025 12:54:02 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE206nnn/GSE206762/soft/GSE206762_family.soft.gz\n",
      "29-Apr-2025 12:54:02 INFO GEOparse - Parsing ./GSE206762_family.soft.gz: \n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SERIES: GSE206762\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - PLATFORM: GPL20795\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261950\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261951\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261952\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261953\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261954\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261955\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261956\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261957\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261958\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261959\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261960\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261961\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261962\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261963\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261964\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261965\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261966\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261967\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261968\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM6261969\n",
      "29-Apr-2025 12:54:02 DEBUG utils - Directory ./ already exists. Skipping.\n",
      "29-Apr-2025 12:54:02 INFO GEOparse - Downloading ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE157nnn/GSE157007/soft/GSE157007_family.soft.gz to ./GSE157007_family.soft.gz\n",
      "100%|██████████| 6.27k/6.27k [00:00<00:00, 153kB/s]\n",
      "29-Apr-2025 12:54:02 DEBUG downloader - Size validation passed\n",
      "29-Apr-2025 12:54:02 DEBUG downloader - Moving /tmp/tmpajj_b0_q to /cis/home/iessien1/Documents/injury_atlas/GSE157007_family.soft.gz\n",
      "29-Apr-2025 12:54:02 DEBUG downloader - Successfully downloaded ftp://ftp.ncbi.nlm.nih.gov/geo/series/GSE157nnn/GSE157007/soft/GSE157007_family.soft.gz\n",
      "29-Apr-2025 12:54:02 INFO GEOparse - Parsing ./GSE157007_family.soft.gz: \n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - DATABASE: GeoMiame\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SERIES: GSE157007\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - PLATFORM: GPL24676\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750298\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750299\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750300\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750301\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750302\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750303\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750304\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750305\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750306\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750307\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750308\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750309\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750310\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750311\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750312\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750313\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750314\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750315\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750316\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750317\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750318\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750319\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750320\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750321\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750322\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750323\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750324\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750325\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750326\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750327\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750328\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750329\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750330\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750331\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750332\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750333\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750334\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750335\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750336\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750337\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750338\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM4750339\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM5684306\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM5684307\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM5684308\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM5684309\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM5684310\n",
      "29-Apr-2025 12:54:02 DEBUG GEOparse - SAMPLE: GSM5684311\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Metadata for GSE206762 ---\n",
      "Title: CircRNA in Frailty\n",
      "Contact Name: David,,Otaegui\n",
      "Contributors:\n",
      "  - David,,Otaegui\n",
      "  - Leire,,Iparraguirre\n",
      "  - Ainhoa,,Alberro\n",
      "  - Itziar,,Vergara\n",
      "  - Ander,,Matheu\n",
      "\n",
      "Expression Type: Expression profiling by high throughput sequencing\n",
      "Overall Design: Frail vs Robust pools analysis\n",
      "\n",
      "Summary:\n",
      "This work describes for the first time a different circular RNA (circRNAs) expression pattern between frail and robust individuals. Moreover, the level of some circRNAs is modulated after a physical intervention. These results suggest that they could be used as minimally invasive biomarkers of frailty. Frailty is an intermediate and reversible geriatric syndrome that often precedes dependency. Therefore, its identification is essential to prevent dependency. Diverse molecules have been proposed as biomarkers of frailty, but none of them has reached clinical practice. Recently, circular RNAs have emerged as new non-coding RNAs. Their regulatory role together with their high stability in biofluids makes them good candidates as biomarkers for several processes, but, to date, no studies have characterized circRNA expression in frailty. We studied RNA from leukocytes of 35 frail and 35 robust individuals. After RNA-Sequencing, circRNA detection was performed by CIRI2 and Circexplorer2 and differential expression analysis by DESeq2. Validation was performed by Quantitative-PCR. A Linear Discriminant Analysis was performed to determine the best circRNA combination to discriminate frails from robusts. Besides, CircRNA candidates were studied in 13 additional elder donors before and after a 3-month physical intervention. We found 89 differentially expressed circRNAs (p-value<0.05, FC>|1.5|) with frailty. The upregulation of hsa_circ_0007817, hsa_circ_0101802 and hsa_circ_0060527 in frail individuals was validated. The combination of hsa_circ_0079284, hsa_circ_0007817 and hsa_circ_0075737 levels showed a great biomarker value with a 95.9% probability of correctly classifying frail and robust individuals. Moreover, hsa_circ_0079284 levels decreased after physical intervention in concordance with an improvement in frailty scores.\n",
      "----------------------------------\n",
      "\n",
      "--- Metadata for GSE157007 ---\n",
      "Title: Multidimensional single-cell analysis of human peripheral blood reveals characteristic features of the immune system landscape in aging and frailty\n",
      "Contact Name: Oscar Junhong,,Luo\n",
      "Contributors:\n",
      "  - Oscar,J,Luo\n",
      "\n",
      "Expression Type: Expression profiling by high throughput sequencing\n",
      "Overall Design: Three cord blood samples, PBMC samples from 3 healthy young, 6 healthy old and 5 frail donors were used for single-cell RNA, TCR and cell surface protein (CCR7, CD45RA, CD4 and CD8) antibody barcoded sequencing to create a single-cell transcriptome and TCR atlas from newborns to frailty.\n",
      "\n",
      "Summary:\n",
      "Frailty is an intermediate status of human ageing process, associated with decompensated homeostasis and death. The immune phenotype of frailty and its underlying cellular and molecular processes remain poorly understood. Here we profiled 114,467 immune cells from cord blood, young adults, healthy and frail elderly using single-cell RNA and TCR V(D)J sequencing. An age-dependent accumulation of cell heterogeneity and transcriptome variability in defined immune cell types was observed. With the specific expression of gene sets, characteristic transcription factors were identified in given immune cell types of certain age group. Trajectory analysis revealed cells from non-frail and frail elderly often fall into distinct trajectories, despite similar chronological age. Numerous TCR clonotypes were shared among T cell subtypes in aged and frail samples, indicating differential pluripotency and resilience capability of aged T cells. Finally, a frailty-specific monocyte subset was identified with exclusively high expression of lncRNAs NEAT1 and MALAT1. Our results discover human frailty-specific immune cell characteristics based on the comprehensive dimensions in immune landscape of ageing and frailty.\n",
      "----------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import GEOparse\n",
    "study = \"GSE236927\"\n",
    "gse_circrna = GEOparse.get_GEO(geo=study)\n",
    "\n",
    "geo_id_pbmc = \"GSE157007\"\n",
    "gse_pbmc = GEOparse.get_GEO(geo=geo_id_pbmc)\n",
    "\n",
    "gse_objects = [gse_circrna, gse_pbmc]\n",
    "geo_ids = [study, geo_id_pbmc]\n",
    "import re # Import regex module for parsing\n",
    "\n",
    "# --- Combine datasets and filter PBMC samples from GSE236927 ---\n",
    "\n",
    "# Identify the GSE object for GSE236927\n",
    "gse_circrna = None\n",
    "target_geo_id = \"GSE236927\"\n",
    "for i, gse in enumerate(gse_objects):\n",
    "    if geo_ids[i] == target_geo_id:\n",
    "        gse_circrna = gse\n",
    "        break\n",
    "\n",
    "if gse_circrna is None:\n",
    "    print(f\"Error: {target_geo_id} not found in the provided gse_objects list.\")\n",
    "    # Handle the error appropriately, e.g., raise an exception or exit\n",
    "    selected_pbmc_samples = {} # Keep variable defined\n",
    "else:\n",
    "    # Define sample numbers to exclude (based on the number in the title like \"PBMC_X\")\n",
    "    # Sample numbers mentioned in the prompt: 2, 8, 14, 15, 18, 21, 28\n",
    "    excluded_sample_numbers = {2, 8, 14, 15, 18, 21, 28}\n",
    "\n",
    "    # Dictionary to store selected GSM objects {gsm_id: gsm_object}\n",
    "    selected_pbmc_samples = {}\n",
    "\n",
    "    print(f\"--- Filtering samples from {target_geo_id} ---\")\n",
    "    print(f\"Excluding sample numbers: {sorted(list(excluded_sample_numbers))}\")\n",
    "\n",
    "    processed_samples = 0\n",
    "    skipped_excluded = 0\n",
    "    skipped_not_pbmc = 0\n",
    "    skipped_no_fi = 0\n",
    "    skipped_parsing_error = 0\n",
    "\n",
    "    # Iterate through samples (GSM objects) in the target GSE\n",
    "    for gsm_id, gsm in gse_circrna.gsms.items():\n",
    "        processed_samples += 1\n",
    "        try:\n",
    "            # --- 1. Extract Sample Number and Check Exclusion ---\n",
    "            title = gsm.metadata.get('title', [''])[0]\n",
    "            sample_num = None\n",
    "            # Use regex to robustly find the number after \"PBMC_\"\n",
    "            match = re.search(r'PBMC_(\\d+)', title, re.IGNORECASE)\n",
    "            if match:\n",
    "                sample_num = int(match.group(1))\n",
    "            else:\n",
    "                # If title format is unexpected, log a warning and skip\n",
    "                # print(f\"Warning: Could not parse sample number from title '{title}' for {gsm_id}. Skipping.\")\n",
    "                skipped_parsing_error += 1\n",
    "                continue\n",
    "\n",
    "            if sample_num in excluded_sample_numbers:\n",
    "                skipped_excluded += 1\n",
    "                continue\n",
    "\n",
    "            # --- 2. Check if Sample Type is PBMC ---\n",
    "            # Check source name or title for PBMC indication\n",
    "            source_name = gsm.metadata.get('source_name_ch1', [''])[0]\n",
    "            is_pbmc = 'pbmc' in source_name.lower() or 'pbmc' in title.lower()\n",
    "\n",
    "            if not is_pbmc:\n",
    "                # print(f\"Skipping {gsm_id} (Sample {sample_num}): Not identified as PBMC (Source: '{source_name}', Title: '{title}').\")\n",
    "                skipped_not_pbmc += 1\n",
    "                continue\n",
    "\n",
    "            # --- 3. Check for Available FI Score ---\n",
    "            characteristics = gsm.metadata.get('characteristics_ch1', [])\n",
    "            fi_score_available = False\n",
    "            fi_score_value = None\n",
    "            for char in characteristics:\n",
    "                # Look for \"frailty index\" or \"FI score\", case-insensitive\n",
    "                if re.search(r'(frailty index|fi score)\\s*:', char, re.IGNORECASE):\n",
    "                    # Extract value after the colon, strip whitespace\n",
    "                    fi_score_str = char.split(':', 1)[-1].strip()\n",
    "                    # Check if the value is meaningful (not empty, 'NA', etc.)\n",
    "                    if fi_score_str and fi_score_str.lower() not in ['na', 'n/a', 'not available', 'unknown']:\n",
    "                        fi_score_available = True\n",
    "                        fi_score_value = fi_score_str # Store the score if needed\n",
    "                        break # Stop searching once a valid FI score is found\n",
    "\n",
    "            if not fi_score_available:\n",
    "                # print(f\"Skipping {gsm_id} (Sample {sample_num}): FI score not available or is NA/empty.\")\n",
    "                skipped_no_fi += 1\n",
    "                continue\n",
    "\n",
    "            # --- If all checks pass, add the sample ---\n",
    "            selected_pbmc_samples[gsm_id] = gsm\n",
    "            # print(f\"Selected {gsm_id} (Sample {sample_num}) - FI Score: {fi_score_value}\")\n",
    "\n",
    "        except ValueError:\n",
    "            # Handle case where sample number extraction fails (e.g., non-integer)\n",
    "            # print(f\"Warning: Could not convert parsed sample number to integer for title '{title}' in {gsm_id}. Skipping.\")\n",
    "            skipped_parsing_error += 1\n",
    "            continue\n",
    "        except Exception as e:\n",
    "            # Catch any other unexpected errors during processing\n",
    "            print(f\"Error processing sample {gsm_id}: {e}. Skipping.\")\n",
    "            skipped_parsing_error += 1\n",
    "            continue\n",
    "\n",
    "    # --- Report Filtering Summary ---\n",
    "    print(f\"\\nFiltering complete for {target_geo_id}:\")\n",
    "    print(f\"  - Total samples processed: {processed_samples}\")\n",
    "    print(f\"  - Skipped (Excluded Number): {skipped_excluded}\")\n",
    "    print(f\"  - Skipped (Not PBMC): {skipped_not_pbmc}\")\n",
    "    print(f\"  - Skipped (FI Score Unavailable): {skipped_no_fi}\")\n",
    "    print(f\"  - Skipped (Parsing/Other Error): {skipped_parsing_error}\")\n",
    "    print(f\"  - Final selected PBMC samples: {len(selected_pbmc_samples)}\")\n",
    "\n",
    "    # Verify the count matches the expected n=88\n",
    "    expected_count = 88\n",
    "    if len(selected_pbmc_samples) == expected_count:\n",
    "        print(f\"Successfully selected the expected {expected_count} samples.\")\n",
    "    else:\n",
    "        print(f\"Warning: Selected {len(selected_pbmc_samples)} samples, but expected {expected_count}.\")\n",
    "\n",
    "    # The 'selected_pbmc_samples' dictionary now contains the filtered GSM objects\n",
    "    # ready for further analysis. Example: Accessing a selected sample's metadata:\n",
    "    # if selected_pbmc_samples:\n",
    "    #     first_gsm_id = list(selected_pbmc_samples.keys())[0]\n",
    "    #     print(f\"\\nMetadata for first selected sample ({first_gsm_id}):\")\n",
    "    #     print(selected_pbmc_samples[first_gsm_id].metadata)\n",
    "\n",
    "# Note: This block replaces the original metadata printing loop.\n",
    "# It focuses specifically on filtering GSE236927 as requested.\n",
    "# The variable `selected_pbmc_samples` holds the result.\n",
    "print() # Add a newline for separation\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "injury",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.22"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
